src/EditorFeatures/Core/Implementation/CommentSelection/ToggleLineCommentCommandHandler.cs

```diff
@@ -0,0 +1,191 @@
+
+        internal async override Task<CommentSelectionResult> CollectEditsAsync(Document document, ICommentSelectionService service,
+            ITextBuffer subjectBuffer, NormalizedSnapshotSpanCollection selectedSpans, ValueTuple command, CancellationToken cancellationToken)
+        {
+            var experimentationService = document.Project.Solution.Workspace.Services.GetRequiredService<IExperimentationService>();
+            if (!experimentationService.IsExperimentEnabled(WellKnownExperimentNames.RoslynToggleLineComment))
+            {
+                return s_emptyCommentSelectionResult;
+            }
+
+            var commentInfo = await service.GetInfoAsync(document, selectedSpans.First().Span.ToTextSpan(), cancellationToken).ConfigureAwait(false);
+            if (commentInfo.SupportsSingleLineComment)
+            {
+                return ToggleLineComment(commentInfo, selectedSpans);
+            }
+
+            return s_emptyCommentSelectionResult;
+        }
+
+        private CommentSelectionResult ToggleLineComment(CommentSelectionInfo commentInfo,
+            NormalizedSnapshotSpanCollection selectedSpans)
+        {
+            var textChanges = ArrayBuilder<TextChange>.GetInstance();
+            var trackingSpans = ArrayBuilder<CommentTrackingSpan>.GetInstance();
+
+            var linesInSelections = selectedSpans.ToDictionary(
+                span => span,
+                span => GetLinesFromSelectedSpan(span).ToImmutableArray());
+
+            Operation operation;
+            // If all the selections are fully commented, uncomment.
+            if (!linesInSelections.Values.Where(lines => SelectionHasUncommentedLines(lines, commentInfo)).Any())
+            {
+                foreach (var selection in linesInSelections)
+                {
+                    UncommentLines(selection.Value, textChanges, trackingSpans, commentInfo);
+                }
+
+                operation = Operation.Uncomment;
+            }
+            else
+            {
+                foreach (var selection in linesInSelections)
+                {
+                    CommentLines(selection.Key, selection.Value, textChanges, trackingSpans, commentInfo);
+                }
+
+                operation = Operation.Comment;
+            }
+
+            return new CommentSelectionResult(textChanges, trackingSpans, operation);
+        }
+
+        private static void UncommentLines(ImmutableArray<ITextSnapshotLine> commentedLines, ArrayBuilder<TextChange> textChanges,
+            ArrayBuilder<CommentTrackingSpan> trackingSpans, CommentSelectionInfo commentInfo)
+        {
+            foreach (var line in commentedLines)
+            {
+                if (!line.IsEmptyOrWhitespace())
+                {
+                    var text = line.GetText();
+                    var commentIndex = text.IndexOf(commentInfo.SingleLineCommentString) + line.Start;
```

Comment:
so, I think we are sure that all lines has singleline comment string once it got here?

Reply:
Yep, all lines are either commented or whitespace.